include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/for_each.hpp

```diff
@@ -0,0 +1,94 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include "../../detail/sycl_utils.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../util.hpp"
+#include "../zip_view.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_range R, typename Fn>
+void
+for_each(ExecutionPolicy&& policy, R&& r, Fn fn)
+{
+    static_assert( // currently only one policy supported
+        std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
+
+    std::vector<sycl::event> events;
```

Comment:
1) if `ExecutionPolicy` is a general policy, not only sycl device policy, we should not use SYCL specific code here. 
2) If the implementation (on current level) works with some "events", a type of event should be general/common type, not SYCL specific. That type may be convertible to sycl::event. For example, see `class __future` in oneDPL. 
3) Due to we know size of ranges::segments(r) before a loop,  it make sense to reserve capacity for `std::vector` before doing `emplace_back` in the loop below.

Reply:
On (1) and (2): since it's in the `sp` subdirectory, everything there can be considered SYCL specific.
We will later need to integrate it better into the general oneDPL design, with policy classes and dispatch tags, etc. But for now this is not the goal.

Reply:
(3) applied

Reply:
good